File: prob3.java
// 232. Implement Queue using Stacks
// S30 Big N Problem #3 {Easy}
// Time Complexity : pop() peek() ~ worst case O(n) but average O(1); push() ~ O(1) 
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
// Two stack implementation
// push the incoming element directly to stack1
// while pop() check the stack2; if it is empty - pop elements from stack1 and fill it in stack2 --- as queue is FIFO
// implement the same strategy for peek()
class MyQueue {

    Stack<Integer> stack1 = new Stack<>();
    Stack<Integer> stack2 = new Stack<>();
    /** Initialize your data structure here. */
    public MyQueue() {
        
    }
    
    /** Push element x to the back of queue. */
    public void push(int x) {
        stack1.push(x);
    }
    
    /** Removes the element from in front of queue and returns that element. */
    public int pop() {
        if(stack2.isEmpty()){
            while(!stack1.isEmpty()){
                stack2.push(stack1.pop());
            }
        }
        return stack2.pop();
    }
    
    /** Get the front element. */
    public int peek() {
        if(stack2.isEmpty()){
            while(!stack1.isEmpty()){
                stack2.push(stack1.pop());
            }
        }
        return stack2.peek();
    }
    
    /** Returns whether the queue is empty. */
    public boolean empty() {
        return stack1.isEmpty() && stack2.isEmpty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */

File: prob3.py
class two_stacks_queue:
    def __init__(self):
        self.s1=[]
        self.s2=[]
    def push(self,val):
        self.s1.append(val)
    def pop(self):
        if (self.s1 == [] and self.s2 == []) :
            return -1
        else:
            if not self.s2:
                while len(self.s1):
                    self.s2.append(self.s1.pop())
                return self.s2.pop()
            else:
                return self.s2.pop()
    def peek(self):
        if not self.s1:
            return self.s2[len(self.s2)-1]
        else:
            while len(self.s1):
                self.s2.append(self.s1.pop())
            return self.s2[len(self.s2)-1]
    def isempty(self):
        return (self.s1 == [] and self.s2 == [])

j= two_stacks_queue()
j.push(10)
j.push(20)
j.push(30)
j.push(40)
j.push(50)
j.push(60)
print(j.pop())
print(j.pop())
print(j.peek())
print(j.pop())










